### ISSUE TITLE
Stack content is not displayed due to invalid memory sections when using gdb stub

### ISSUE THREAD

**Issue by boaster** [hugsy/gef#14]
I also have another issue. It looks like the stack is not being properly interpreted as the correct data type.

```bash
------------------------------------------------------------------------------------------[regs]
$zero       0x00000000 $at         0xfffffff8 $v0         0x76fff5e4 $v1         0x0000002c 
$a0         0x76fff5f4 $a1         0x00000000 $a2         0x00000000 $a3         0x76fff5f4 
$t0         0x00000000 $t1         0x00000000 $t2         0x00000000 $t3         0x00000200 
$t4         0x00000100 $t5         0xfffffffe $t6         0x00000001 $t7         0x004012f8 
$s0         0x56153a1d $s1         0xffffffff $s2         0x76fff6e4 $s3         0x00400c38 
$s4         0x00000001 $s5         0x00400e80 $s6         0x00000000 $s7         0x00000000 
$t8         0x00000024 $t9         0x766ee1d0 $k0         0x00000000 $k1         0x00000000 
$s8         0x76fff5c0 $sr         0x20000010 $bad        0x00000000 $cause      0x00000000 
$pc         0x004012fc $sp         0x76fff5c0 $hi         0x00000249 $lo         0x0001cbed 
$fir        0x00739300 $fsr        0x00000000 $ra         0x004012f8 $gp         0x004520e0 

------------------------------------------------------------------------------------------[stack]
0x76fff5c0|+0x00: 		← $sp
0x76fff5c4|+0x04: 
0x76fff5c8|+0x08: 
0x76fff5cc|+0x0c: 
0x76fff5d0|+0x10: 
0x76fff5d4|+0x14: 
0x76fff5d8|+0x18: 
0x76fff5dc|+0x1c: 
0x76fff5e0|+0x20: 
0x76fff5e4|+0x24: 
------------------------------------------------------------------------------------------[code:mips:3000]
0x4012f0	 <main+1136>  jalr t9
0x4012f4	 <main+1140>  nop
0x4012f8	 <main+1144>  lw gp,16(s8)
0x4012fc	 <main+1148>  li v0,2 		 ← $pc
0x401300	 <main+1152>  sh v0,36(s8)
0x401304	 <main+1156>  lw v0,-32744(gp)
0x401308	 <main+1160>  nop
0x40130c	 <main+1164>  addiu v0,v0,-21780
0x401310	 <main+1168>  lhu v0,4(v0)
0x401314	 <main+1172>  nop
0x401318	 <main+1176>  move a0,v0
0x40131c	 <main+1180>  lw t9,-32672(gp)
------------------------------------------------------------------------------------------[trace]
#0  0x004012fc in main ()
0x004012fc in main ()
gef> x/10wx $sp
0x76fff5c0:	0x0044a868	0x00000032	0x0044a868	0xfefefeff
0x76fff5d0:	0x004520e0	0x00400c38	0x00000001	0x00400e80
0x76fff5e0:	0x00000000	0x00000000
gef> 
```

So far from my debugging I see that context_stack() calls inspect_stack(). Inspect_stack() calls _do_inspect_stack(), there line 3305 variable 'l' does not add the dereferenced stack pointers.

```python
3305             l += sep.join(addrs[1:])
```
Looks like dereference_from() calls lookup_address(). There at line 1073, the sect variable is None due to get_process_maps() uses get_pid() and it reads the debugged process's memory map from proc.

```python
 920         pid = get_pid()
 921         f = open('/proc/%d/maps' % pid)
```
I think the issue is that get_pid() does not return a valid process ID on the host system because my process is running inside qemu and it does not exist as a process on it's own.

Here is an example where I print the returned pid and the inferior object.
```python
 920         pid = get_pid()
 921         print(pid)
 922         f = open('/proc/%d/maps' % pid)
```

```bash
gef> target remote localhost:12345
Remote debugging using localhost:12345
Reading symbols from /lib/ld-uClibc.so.0...(no debugging symbols found)...done.
Loaded symbols for /lib/ld-uClibc.so.0
<gdb.Inferior object at 0x7fb5cbea7288>
42000
------------------------------------------------------------------------------------------[regs]
$zero       0x00000000 $at         0x00000000 $v0         0x00000000 $v1         0x00000000 
$a0         0x00000000 $a1         0x00000000 $a2         0x00000000 $a3         0x00000000 
$t0         0x00000000 $t1         0x00000000 $t2         0x00000000 $t3         0x00000000 
$t4         0x00000000 $t5         0x00000000 $t6         0x00000000 $t7         0x00000000 
$s0         0x00000000 $s1         0x00000000 $s2         0x00000000 $s3         0x00000000 
$s4         0x00000000 $s5         0x00000000 $s6         0x00000000 $s7         0x00000000 
$t8         0x00000000 $t9         0x00000000 $k0         0x00000000 $k1         0x00000000 
$s8         0x00000000 $sr         0x20000010 $bad        0x00000000 $cause      0x00000000 
$pc         0x767b8ab0 $sp         0x76fff6e0 $hi         0x00000000 $lo         0x00000000 
$fir        0x00739300 $fsr        0x00000000 $ra         0x00000000 $gp         0x00000000 

------------------------------------------------------------------------------------------[stack]
<gdb.Inferior object at 0x7fb5cbea7288>
42000
<gdb.Inferior object at 0x7fb5cbea7288>
42000
0x76fff6e0|+0x00: 		← $sp
<gdb.Inferior object at 0x7fb5cbea7288>
42000
<gdb.Inferior object at 0x7fb5cbea7288>
42000
0x76fff6e4|+0x04: 
<gdb.Inferior object at 0x7fb5cbea7288>
42000
<gdb.Inferior object at 0x7fb5cbea7288>
42000
0x76fff6e8|+0x08: 
<gdb.Inferior object at 0x7fb5cbea7288>
....
```
Here is the actual processes:

```bash
➜  ~  ps aux | grep mips | grep -v grep
user     53565  0.0  1.1  92360 24464 pts/0    S+   15:23   0:00 gdb-multiarch -q ./mips
user     53630 96.5  0.1 2055184 2568 pts/7    R+   15:46   1:59 qemu-mips -g 12345 ./mips
➜  ~  
```

The process qemu-mips, is the process whose memory map we need.

```bash
➜  ~  cat /proc/53630/maps       
7ffbee60b000-7ffbeea0b000 ---p 00000000 00:00 0 
7ffbeea0b000-7ffbeea15000 r-xp 00000000 08:01 656230                     /home/user/openwrt/mips
7ffbeea15000-7ffbeea55000 ---p 00000000 00:00 0 
7ffbeea55000-7ffbeea56000 rw-p 0000a000 08:01 656230                     /home/user/openwrt/mips
7ffbeea56000-7ffc64dc3000 ---p 00000000 00:00 0 
7ffc64dc3000-7ffc64dc9000 r-xp 00000000 08:01 665474                     /home/user/openwrt/cross-compiler-mips/lib/ld-uClibc-0.9.30.1.so
7ffc64dc9000-7ffc64e08000 ---p 00000000 00:00 0 
7ffc64e08000-7ffc64e0a000 rw-p 00005000 08:01 665474                     /home/user/openwrt/cross-compiler-mips/lib/ld-uClibc-0.9.30.1.so
7ffc64e0a000-7ffc64e0b000 ---p 00000000 00:00 0 
7ffc64e0b000-7ffc68a0b000 rw-p 00000000 00:00 0 
7ffc68a0b000-7ffc68a48000 r-xp 00000000 08:01 921979                     /lib/x86_64-linux-gnu/libpcre.so.3.13.1
7ffc68a48000-7ffc68c47000 ---p 0003d000 08:01 921979                     /lib/x86_64-linux-gnu/libpcre.so.3.13.1
7ffc68c47000-7ffc68c48000 r--p 0003c000 08:01 921979                     /lib/x86_64-linux-gnu/libpcre.so.3.13.1
7ffc68c48000-7ffc68c49000 rw-p 0003d000 08:01 921979                     /lib/x86_64-linux-gnu/libpcre.so.3.13.1
7ffc68c49000-7ffc68e04000 r-xp 00000000 08:01 923688                     /lib/x86_64-linux-gnu/libc-2.19.so
7ffc68e04000-7ffc69003000 ---p 001bb000 08:01 923688                     /lib/x86_64-linux-gnu/libc-2.19.so
7ffc69003000-7ffc69007000 r--p 001ba000 08:01 923688                     /lib/x86_64-linux-gnu/libc-2.19.so
7ffc69007000-7ffc69009000 rw-p 001be000 08:01 923688                     /lib/x86_64-linux-gnu/libc-2.19.so
7ffc69009000-7ffc6900e000 rw-p 00000000 00:00 0 
7ffc6900e000-7ffc69027000 r-xp 00000000 08:01 923689                     /lib/x86_64-linux-gnu/libpthread-2.19.so
7ffc69027000-7ffc69226000 ---p 00019000 08:01 923689                     /lib/x86_64-linux-gnu/libpthread-2.19.so
7ffc69226000-7ffc69227000 r--p 00018000 08:01 923689                     /lib/x86_64-linux-gnu/libpthread-2.19.so
7ffc69227000-7ffc69228000 rw-p 00019000 08:01 923689                     /lib/x86_64-linux-gnu/libpthread-2.19.so
7ffc69228000-7ffc6922c000 rw-p 00000000 00:00 0 
7ffc6922c000-7ffc69331000 r-xp 00000000 08:01 922568                     /lib/x86_64-linux-gnu/libm-2.19.so
7ffc69331000-7ffc69530000 ---p 00105000 08:01 922568                     /lib/x86_64-linux-gnu/libm-2.19.so
7ffc69530000-7ffc69531000 r--p 00104000 08:01 922568                     /lib/x86_64-linux-gnu/libm-2.19.so
7ffc69531000-7ffc69532000 rw-p 00105000 08:01 922568                     /lib/x86_64-linux-gnu/libm-2.19.so
7ffc69532000-7ffc69638000 r-xp 00000000 08:01 922562                     /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0
7ffc69638000-7ffc69837000 ---p 00106000 08:01 922562                     /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0
7ffc69837000-7ffc69838000 r--p 00105000 08:01 922562                     /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0
7ffc69838000-7ffc69839000 rw-p 00106000 08:01 922562                     /lib/x86_64-linux-gnu/libglib-2.0.so.0.4002.0
7ffc69839000-7ffc6983a000 rw-p 00000000 00:00 0 
7ffc6983a000-7ffc69841000 r-xp 00000000 08:01 923691                     /lib/x86_64-linux-gnu/librt-2.19.so
7ffc69841000-7ffc69a40000 ---p 00007000 08:01 923691                     /lib/x86_64-linux-gnu/librt-2.19.so
7ffc69a40000-7ffc69a41000 r--p 00006000 08:01 923691                     /lib/x86_64-linux-gnu/librt-2.19.so
7ffc69a41000-7ffc69a42000 rw-p 00007000 08:01 923691                     /lib/x86_64-linux-gnu/librt-2.19.so
7ffc69a42000-7ffc69a65000 r-xp 00000000 08:01 923682                     /lib/x86_64-linux-gnu/ld-2.19.so
7ffc69c13000-7ffc69c47000 rw-p 00000000 00:00 0 
7ffc69c4a000-7ffc69c64000 rw-p 00000000 00:00 0 
7ffc69c64000-7ffc69c65000 r--p 00022000 08:01 923682                     /lib/x86_64-linux-gnu/ld-2.19.so
7ffc69c65000-7ffc69c66000 rw-p 00023000 08:01 923682                     /lib/x86_64-linux-gnu/ld-2.19.so
7ffc69c66000-7ffc69c67000 rw-p 00000000 00:00 0 
7ffc69c67000-7ffc69dd2000 r-xp 00000000 08:01 917736                     /usr/bin/qemu-mips
7ffc69fd2000-7ffc6a012000 r--p 0016b000 08:01 917736                     /usr/bin/qemu-mips
7ffc6a012000-7ffc6a019000 rw-p 001ab000 08:01 917736                     /usr/bin/qemu-mips
7ffc6a019000-7ffc6a077000 rw-p 00000000 00:00 0 
7ffc6a077000-7ffc6c078000 rwxp 00000000 00:00 0 
7ffc6c078000-7ffc6c07b000 rw-p 00000000 00:00 0 
7ffc6c27f000-7ffc6c2a0000 rw-p 00000000 00:00 0                          [heap]
7ffff98f0000-7ffff9911000 rw-p 00000000 00:00 0                          [stack]
7ffff99fe000-7ffff9a00000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
➜  ~  
```

I don't know how to fix that yet, maybe u can help me with some suggestions. Two solutions I'm thinking of are:
1. Maybe try a different method of supplying the correct PID (althought I'm thinking this is more like a hack solution and not a real solution to the problem)
2. Use a different method to read the memory instead of reading the proc filesystem (this looks like a better more reliable long term solution)

**Comment by hugsy** [comment]
Hi,

Thanks for this other bug. FYI, support for remote debug is still experimental in `gef` see ( #7 ) because as you may have noticed, there are a lot of constraints to overcome, which you do not have when debugging locally. I'm trying to do that in a clean way so it may take time, but suggestions, bugs and PR are sure welcome.

As you noticed, one of the big issues here is to get as many info from the process memory layout during runtime, for which I used `/proc/self/maps` because `maintenance info files`  and `info files` do not provide enough information.

Even though I'm not looking for high perf, reading the memory from `gdb` is a perf killer, but you are right as I believe it is the right way to go. Happy to discuss if you have ideas about it.

Cheers

**Comment by boaster** [comment]
I know it's quick and dirty implementation but I will go with this for now...

inside dereference_from()

```python
2796             if addr is None:
2797                 msg.append( "%#x" % ( long(deref) ))
2798                 value_from_stack = long(DereferenceCommand.dereference(value))
2799                 msg.append( format_address(value_from_stack) )
2800                 examine_string = gdb.execute("x/s %#x" % value_from_stack, to_string=True).strip().replace("\n", "").replace("\r", "")
2801                 stack_string = re.findall('"(.*)"', examine_string)
2802                 if stack_string:
2803                     if '\\' not in stack_string[0]:
2804                         msg.append(stack_string[0])
2805                 break
2806             msg.append( "%s" % format_address( long(deref) ))
```

```bash
gef> c
Continuing.
------------------------------------------------------------------------------------------[regs]
$zero       0x00000000 $at         0xfffffff8 $v0         0x76fff5e4 $v1         0x0000002c 
$a0         0x76fff5f4 $a1         0x00000000 $a2         0x00000000 $a3         0x76fff5f4 
$t0         0x00000000 $t1         0x00000000 $t2         0x00000000 $t3         0x00000200 
$t4         0x00000100 $t5         0xfffffffe $t6         0x00000001 $t7         0x004012f8 
$s0         0x5615c73e $s1         0xffffffff $s2         0x76fff6e4 $s3         0x00400c38 
$s4         0x00000001 $s5         0x00400e80 $s6         0x00000000 $s7         0x00000000 
$t8         0x00000024 $t9         0x766ee1d0 $k0         0x00000000 $k1         0x00000000 
$s8         0x76fff5c0 $sr         0x20000010 $bad        0x00000000 $cause      0x00000000 
$pc         0x004012f8 $sp         0x76fff5c0 $hi         0x00000249 $lo         0x0001cbed 
$fir        0x00739300 $fsr        0x00000000 $ra         0x004012f8 $gp         0x76806010 

------------------------------------------------------------------------------------------[stack]
0x76fff5c0|+0x00: 0x0044a868 → ubuntu		← $sp
0x76fff5c4|+0x04: 0x00000032
0x76fff5c8|+0x08: 0x0044a868 → ubuntu
0x76fff5cc|+0x0c: 0xfefefeff
0x76fff5d0|+0x10: 0x004520e0
0x76fff5d4|+0x14: 0x00400c38
0x76fff5d8|+0x18: 0x00000001
0x76fff5dc|+0x1c: 0x00400e80
0x76fff5e0|+0x20: 0x00000000
0x76fff5e4|+0x24: 0x00000000
------------------------------------------------------------------------------------------[code:mips:3000]
0x4012ec	 <main+1132>  nop
0x4012f0	 <main+1136>  jalr t9
0x4012f4	 <main+1140>  nop
0x4012f8	 <main+1144>  lw gp,16(s8) 		 ← $pc
0x4012fc	 <main+1148>  li v0,2
0x401300	 <main+1152>  sh v0,36(s8)
0x401304	 <main+1156>  lw v0,-32744(gp)
0x401308	 <main+1160>  nop
0x40130c	 <main+1164>  addiu v0,v0,-21780
0x401310	 <main+1168>  lhu v0,4(v0)
0x401314	 <main+1172>  nop
0x401318	 <main+1176>  move a0,v0
------------------------------------------------------------------------------------------[trace]
#0  0x004012f8 in main ()

Breakpoint 2, 0x004012f8 in main ()
gef> x/10wx $sp
0x76fff5c0:	0x0044a868	0x00000032	0x0044a868	0xfefefeff
0x76fff5d0:	0x004520e0	0x00400c38	0x00000001	0x00400e80
0x76fff5e0:	0x00000000	0x00000000
gef> 
```

I would say at least implement something like line 2799 so when the stack is printed out, it will at least show the content of the 10 words on it in hex...

**Comment by hugsy** [comment]
This bug should be fixed in a cleaner way by commit aabc740.
The command `gef-remote` allows to grasp all remote information and access them from `gef` as if they were local.
